DiskSetup.sh

```diff
@@ -50,51 +88,61 @@ function CarveLVM() {
    parted -s "${CHROOTDEV}" -- mklabel msdos mkpart primary "${FSTYPE}" 2048s ${BOOTDEVSZ} \
       mkpart primary "${FSTYPE}" ${BOOTDEVSZ} 100% set 2 lvm
 
+   # Gather info to diagnose seeming /boot race condition
+   if [[ $(grep -q "${BOOTLABEL}" /proc/mounts)$? -eq 0 ]]
+   then
+     tail -n 100 /var/log/messages
+     sleep 3
+   fi
+
    # Stop/umount boot device, in case parted/udev/systemd managed to remount it
-  systemctl stop boot.mount || true
+   # again.
+   systemctl stop boot.mount || true
+
+   # Create /boot filesystem
+   mkfs -t "${FSTYPE}" "${MKFSFORCEOPT}" -L "${BOOTLABEL}" \
+     "${CHROOTDEV}${PARTPRE}1" || \
+       err_exit "Failure creating filesystem - /boot"
+
+   ## Create LVM objects
 
-   # Create LVM objects
-   LVCSTAT=0
    # Let's only attempt this if we're a secondary EBS
    if [[ ${CHROOTDEV} == /dev/xvda ]] || [[ ${CHROOTDEV} == /dev/nvme0n1 ]]
    then
       echo "Skipping explicit pvcreate opertion... " 
    else
       pvcreate "${CHROOTDEV}${PARTPRE}2" || LogBrk 5 "PV creation failed. Aborting!"
    fi
+
+   # Create root VolumeGroup
    vgcreate -y "${VGNAME}" "${CHROOTDEV}${PARTPRE}2" || LogBrk 5 "VG creation failed. Aborting!"
-   lvcreate --yes -W y -L "${ROOTVOL[1]}" -n "${ROOTVOL[0]}" "${VGNAME}" || LVCSTAT=1
-   lvcreate --yes -W y -L "${SWAPVOL[1]}" -n "${SWAPVOL[0]}" "${VGNAME}" || LVCSTAT=1
-   lvcreate --yes -W y -L "${HOMEVOL[1]}" -n "${HOMEVOL[0]}" "${VGNAME}" || LVCSTAT=1
-   lvcreate --yes -W y -L "${VARVOL[1]}" -n "${VARVOL[0]}" "${VGNAME}" || LVCSTAT=1
-   lvcreate --yes -W y -L "${LOGVOL[1]}" -n "${LOGVOL[0]}" "${VGNAME}" || LVCSTAT=1
-   lvcreate --yes -W y -l "${AUDVOL[1]}" -n "${AUDVOL[0]}" "${VGNAME}" || LVCSTAT=1
 
-   if [[ ${LVCSTAT} = 1 ]]
-   then
-      echo "Failed creating one or more volumes. Aborting"
-      exit 1
-   fi
+   # Create LVM2 volume-objects by iterating ${PARTITIONARRAY}
+   for ARRAYELEM in ${PARTITIONARRAY[*]}
+   do
+      MOUNTPT="$( cut -d ':' -f 1 <<< "${ARRAYELEM}")"
+      VOLNAME="$( cut -d ':' -f 2 <<< "${ARRAYELEM}")"
+      VOLSIZE="$( cut -d ':' -f 3 <<< "${ARRAYELEM}")"
 
-   # Gather info to diagnose seeming /boot race condition
-   if [[ $(grep -q "${BOOTLABEL}" /proc/mounts)$? -eq 0 ]]
-   then
-     tail -n 100 /var/log/messages
-     sleep 3
-   fi
-
-   # Stop/umount boot device, in case parted/udev/systemd managed to remount it
-   # again.
-  systemctl stop boot.mount || true
+      # Create LVs
+      if [[ ${VOLSIZE} =~ FREE ]]
+      then
+         lvcreate --yes -W y -l "100%FREE" -n "${VOLNAME}" "${VGNAME}" || \
```

Comment:
could reduce some of the duplication by just setting VOLSIZE in the if block, and calling lvcreate once afterwards...

```suggestion
         VOLSIZE="100%FREE"
```

Reply:
Interface is making the suggestion difficult to parse. That said, was originally trying to accomplish two things:

* Leave the VOLSIZE-specification flexible (i.e., whether someone specifies the generic token `FREE` or the LVM-specific token `100%FREE`).
* Account for LVM2's requirements with respect to size-specification. Which is to say, allowing one to use the `%FREE` size-spec requires the use of the `-l`  flag while `-L` is required when using sizes specified in units other than extents.

Reply:
Updated to use:

~~~
# Create LVs
if [[ ${VOLSIZE} =~ FREE ]]
then
   VOLFLAG="-l"
   VOLSIZE="100%FREE"
else
   VOLFLAG="-L"
   VOLSIZE+="g"
fi
lvcreate --yes -W y "${VOLFLAG}" "${VOLSIZE}" -n "${VOLNAME}" "${VGNAME}" || \
  err_exit "Failure creating LVM2 volume '${VOLNAME}'"
~~~